connectors/sources/gmail.py

```diff
@@ -0,0 +1,352 @@
+#
+# Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+# or more contributor license agreements. Licensed under the Elastic License 2.0;
+# you may not use this file except in compliance with the Elastic License 2.0.
+#
+import json
+from functools import cached_property
+
+import fastjsonschema
+from fastjsonschema import JsonSchemaValueException
+
+from connectors.filtering.validation import (
+    AdvancedRulesValidator,
+    SyncRuleValidationResult,
+)
+from connectors.source import BaseDataSource, ConfigurableFieldValueError
+from connectors.sources.google import (
+    GMailClient,
+    GoogleDirectoryClient,
+    MessageFields,
+    UserFields,
+)
+from connectors.utils import base64url_to_base64, iso_utc
+
+ACCESS_CONTROL = "_allow_access_control"
+
+GMAIL_API_TIMEOUT = GOOGLE_DIRECTORY_TIMEOUT = 1 * 60  # 1 min
+
+
+class GMailAdvancedRulesValidator(AdvancedRulesValidator):
+    MESSAGES_SCHEMA_DEFINITION = {
+        "type": "array",
+        "items": {"type": "string"},
+        "minItems": 1,
+    }
+
+    SCHEMA_DEFINITION = {
+        "type": "object",
+        "properties": {"messages": MESSAGES_SCHEMA_DEFINITION},
+        "additionalProperties": False,
+    }
+
+    SCHEMA = fastjsonschema.compile(
+        definition=SCHEMA_DEFINITION,
+    )
+
+    async def validate(self, advanced_rules):
+        if len(advanced_rules) == 0:
+            return SyncRuleValidationResult.valid_result(
+                SyncRuleValidationResult.ADVANCED_RULES
+            )
+
+        try:
+            GMailAdvancedRulesValidator.SCHEMA(advanced_rules)
+
+            return SyncRuleValidationResult.valid_result(
+                rule_id=SyncRuleValidationResult.ADVANCED_RULES
+            )
+        except JsonSchemaValueException as e:
+            return SyncRuleValidationResult(
+                rule_id=SyncRuleValidationResult.ADVANCED_RULES,
+                is_valid=False,
+                validation_message=e.message,
+            )
+
+
+def _message_doc(message):
```

Comment:
Why a function starts with `_`?

Reply:
I've prefixed it with `_` to indicate that this function is module-private.